---
title: "Taxonomic data transformation"
output: html_notebook
---

# Taxonomic data transformation

## Reduction into a data.frame with relative abundances for each clade

First we load the data into a data.frame called **raw_df**.

```{r}
raw_df = read.csv(file="../raw_data/quantification_by_contig_lineage_all.csv",
                  sep="\t",
                  na.strings = "None")
head(raw_df)
```

We will remove some columns and rename others to make it easier to work with. First, we create a new data.frame that will contain all phylogenetic levels in distinct columns rather than in the same column as in raw_data.

```{r}
# strsplit split strings of the lineage_by_level using the "; " pattern
splitted_levels = strsplit(raw_df$lineage_by_level, "; ")
abundance_df = t(data.frame(splitted_levels))

# Renaming the columns according to levels
colnames(abundance_df) = c("kingdom", "phylum", "class", "order", "familly", "genre", "species")

# Removing complicated row names
#row.names(abundance_df) = NULL

# Removing temporary variables
#rm(splitted_levels)

head(abundance_df)
```

Then we will add the abundance values for each row. Each sample will be represented in a different column. For example the column *abund1* stores abundances of the first sample. The relative abundance is computed by dividing the number of reads by the total number of reads for the given sample.

```{r}
# We go through all combinations of columns name's containing the number of reads. # Then, we add a column in abundance_df containing relative abundances. 
for (sample_number in 1:34){
  if (sample_number < 10){
    col_name=paste("nb_reads_Holopig_colon0",sample_number,"_quantif_percontig", sep="")
  }
  else {
    col_name=paste("nb_reads_Holopig_colon",sample_number,"_quantif_percontig", sep="")
  }
  abundance_df[paste("abond",sample_number,sep="")] = raw_df[col_name]/sum(raw_df[col_name])
}

# Removing unnecessary variables
rm(list= c("col_name", "sample_number"))
```