---
title: "Taxonomic data preprocessing"
output: html_notebook
---

# Reduction into a data.frame with relative abundances for each clade

This part aims to treat the result of metagenomic sequencing by reducing a data sheet of quantity and quality of reads for each individual (columns) per taxon (lines) in a data frame containing for one taxon and one individual (lines) the different taxonomic information and the relative abundance of reads.

## Importing the raw dataset

First we load the data into a data.frame called **raw_df**.

```{r}
raw_df = read.csv(file="../raw_data/quantification_by_contig_lineage_all.tsv",
                  sep="\t",
                  na.strings = "None")
head(raw_df)
```

We will remove some columns and rename others to make it easier to work with. First, we create a new data.frame that will contain all phylogenic levels in distinct columns rather than in the same column as in raw_data.

```{r}
# strsplit split strings of the lineage_by_level using the "; " pattern
splitted_levels = strsplit(raw_df$lineage_by_level, "; ")
abundance_df = as.data.frame( t(data.frame(splitted_levels)) )

# Renaming the columns according to taxonomic ranks
colnames(abundance_df) = c("kingdom", "phylum", "class", "order", "family", "genus", "species")

# Removing complicated row names
rownames(abundance_df) = NULL

# Removing temporary variables
rm(splitted_levels)

head(abundance_df)
```

Then we will add the abundance values for each row. Each sample will be represented in a different column. For example the column *abund1* stores abundances of the first sample. The relative abundance is computed by dividing the number of reads by the total number of reads for the given sample.

```{r}
# We go through all combinations of columns name's containing the number of reads. # Then, we add a column in abundance_df containing relative abundances. 
for (sample_number in 1:34){
  if (sample_number < 10){
    col_name=paste("nb_reads_Holopig_colon0",sample_number,"_quantif_percontig", sep="")
  }
  else {
    col_name=paste("nb_reads_Holopig_colon",sample_number,"_quantif_percontig", sep="")
  }
  abundance_df[paste("abond",sample_number,sep="")] = raw_df[col_name]/sum(raw_df[col_name])
}

# Removing unnecessary variables
rm(list= c("col_name", "sample_number"))
```

We will add the *root* column for the phylogenic tree.

```{r}
abundance_df$root = as.factor( rep("root", dim(abundance_df)[1]) )

# Reordering columns
abundance_df = abundance_df[, c(dim(abundance_df)[2], 1:(dim(abundance_df)[2]-1))]

head(abundance_df)
```

## Filtering rows according to the relative abundance mean

```{r}
unfiltered_abundance = abundance_df
abundance_df$Mean = rowMeans(abundance_df[9:42])
```

In an attempt to remove less valuable values, we will filter them according to their mean. The mean correspond to the relative abundance mean across samples for each row. In order to do that, we compute the quantile associated to a 90% probability (i.e. the value at which 90% of observation are included). This threshold divide the 10% best rows from the 90% others.

```{r}
## Getting the quantile as threshold
thres = quantile(abundance_df$Mean, prob=c(.90), type=1)

library(ggplot2)
ggplot(abundance_df, aes(x=Mean)) + 
  geom_histogram(binwidth=0.01) +
  geom_vline(xintercept = thres, linetype = "dashed", linewidth = 1.5, 
             color = "red")+
  geom_label(label = paste0("Threshold\n",round(thres, 3)), x = thres, y = 100, 
             color = "red")+
  xlab("Relative abundance mean")+
  ggtitle("Ditribution of relative abundance means")
```

The next cell remove the rows where the mean relative abundance is less than the threshold.

```{r}
abundance_df = subset(abundance_df, Mean >= thres)
```

## Transforming the data.frame for ggplot

We got a final manipulation to do. In order to make the heatmap and the boxplot, the category needs to be a factor (1 column). So the abundance needs to be aggregated for each row and each sample which will lead to a 2649\*34 = 90066 rows (rows times number of samples).

```{r}
base_taxo = abundance_df[, 1:8]
## defining the group of each individual
category = c("Control", "Colistin", "Control", "Control", "Control", "Control",
              "Colistin", "Colistin", "Control", "Colistin", "Colistin", "Control",
              "Control", "Control", "Control", "Control", "Control", "Colistin",
              "Colistin", "Colistin", "Colistin", "Colistin", "Control", "Control",
              "Control", "Control", "Control", "Colistin", "Control", "Colistin",
              "Colistin", "Colistin", "Colistin", "Colistin")

final_df = data.frame()
for (sample in 1:34){
# identification of the taxon
  tmp = base_taxo
# identification of the individual
  tmp$ind = as.factor(rep(sample, dim(base_taxo)[1]))
# identification of the group
  tmp$cat = as.factor(rep(category[sample], dim(base_taxo)[1]))
# relative abundance of the species if the individual
  tmp$abond = abundance_df[[8+sample]]
# adding the data of the individual to the final dataframe
  final_df = rbind(final_df, tmp)
}

rm(list= c("tmp", "base_taxo", "category", "sample"))
```

Finally we can export both *final_df* and *abundance_df*.

```{r}
write.table(final_df, 
            file = "../refined_data/filtered_taxonomic_abundance.csv",
            col.names = TRUE,
            row.names = FALSE,
            sep=";")

write.table(abundance_df, 
            file = "../refined_data/filtered_taxonomy_X_sample.csv",
            col.names = TRUE,
            row.names = FALSE,
            sep=";")

write.table(unfiltered_abundance, 
            file = "../refined_data/complete_taxonomy_X_sample.csv",
            col.names = TRUE,
            row.names = FALSE,
            sep=";")

```
